File: rag_finalised/realtime/pusher_client.py
import os
from typing import Optional

from pusher import Pusher


def _pusher_enabled() -> bool:
    """
    Controls whether Pusher is active via env var:
    PUSHER=1/true/yes/on  -> enabled
    anything else         -> disabled
    """
    value = os.getenv("PUSHER", "").strip().lower()
    return value in {"1", "true", "yes", "on"}


def get_pusher() -> Optional[Pusher]:
    """
    Returns a configured Pusher client or None when:
    - PUSHER flag is disabled, or
    - any of the required env vars are missing.
    """
    if not _pusher_enabled():
        return None

    app_id = os.getenv("PUSHER_APP_ID")
    key = os.getenv("PUSHER_KEY")
    secret = os.getenv("PUSHER_SECRET")
    cluster = os.getenv("PUSHER_CLUSTER")

    if not all([app_id, key, secret, cluster]):
        return None

    return Pusher(
        app_id=app_id,
        key=key,
        secret=secret,
        cluster=cluster,
        ssl=True,
    )


def expert_channel(name: str) -> str:
    """
    Builds a channel name for an expert, normalized to allowed characters.
    """
    base = name.lower()
    safe = []
    for ch in base:
        if ch.isalnum():
            safe.append(ch)
        elif ch in {" ", "-", "_"}:
            safe.append("-")
    slug = "".join(safe).strip("-")
    return f"expert-chat-{slug or 'unknown'}"

File: rag_frontend/src/realtime/pusherClient.ts
import Pusher from 'pusher-js'

const key = import.meta.env.VITE_PUSHER_KEY
const cluster = import.meta.env.VITE_PUSHER_CLUSTER

export function createPusherClient() {
  if (!key || !cluster) {
    return null
  }

  return new Pusher(key, {
    cluster,
    forceTLS: true,
  })
}

export function expertChannelName(expertName: string) {
  const base = expertName.toLowerCase()
  const safe = base
    .split('')
    .map((ch) => {
      if (/[a-z0-9]/.test(ch)) return ch
      if (ch === ' ' || ch === '-' || ch === '_') return '-'
      return ''
    })
    .join('')
    .replace(/-+/g, '-')
    .replace(/^-|-$/g, '')

  return `expert-chat-${safe || 'unknown'}`
}

